### DISCUSSION TITLE
#[Url(history: true)] when updating multiple properties creates multiple (redundant) history steps

### DISCUSSION THREAD

**Discussion by goldmerc** [livewire/livewire#8640] [7. Bugs (archived)]
### Livewire version

v3.5.2

### Laravel version

v11.15.0

### Which PHP version are you using?

PHP 8.3

### Steps To Reproduce

If I update two properties in the same request, both with #[Url(history: true)] then livewire creates two browser history entries rather than one. I've attached a repo that shows the issue but for quick reference..

For example -

```
class TestQueryString extends Component
{
    #[Url(history: true)]
    public $minPrice;

    #[Url(history: true)]
    public $maxPrice;
}
```
```
<form wire:submit="$refresh">
    <input wire:model="minPrice"/>
    <input wire:model="maxPrice"/>
    <button type="submit">Update Price Range</button>
</form>
```
If I enter min price 20 and max price 50 and submit the form, I see the expected url -

`app.test/price-range-test?minPrice=20&maxPrice=50`

However, when I hit back, I get -

`app.test/price-range-test?minPrice=20`

But I would expect to go directly back to -

`app.test/price-range-test`

Is this the expected behaviour? Seems like a bug to me but maybe I'm missing something?

Thanks in advance,

Jonathan

### Repository or Wirebox URL

https://github.com/goldmerc/livewire-query-string

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by goldmerc** [comment]
Been testing further, I tried using the queryString() function rather than the Url attribute. I think I've found another bug. If you use the querystring function it automatically uses history even without setting it...
```
    public $minPrice = '';
    
    public $maxPrice = '';

    protected function queryString()
    {
        return [
            'minPrice',
            'maxPrice'
        ];
    }
```
The above code Is enough to get history steps, even though history => true has not been set.

**Comment by joshhanley** [comment]
@goldmerc I have replied on discussion #8636. Please don't open two discussions for the same thing. Thanks!

**Comment by goldmerc** [comment]
Hi Josh, thanks for your help and sorry for the duplication. After I opened the previous thread, I did more testing and convinced myself this issue is a bug. So, I opened this bug report with the necessary repo to recreate the issue. I should have deleted the other thread but it didn't occur to me.

As this thread has the repo attached, is it better to continue our discussion here? What's your preference?

Thanks again for your help.

**Reply by joshhanley** [reply]
@goldmerc no worries! In future, you should just be able to change the category of the discussion from help to bug (or if not tag a maintainer and they can), so you don't need to raise a second discussion 🙂

I will copy my answer over from the other discussion below, so we can continue the discussion here.

**Comment by joshhanley** [comment]
Copied from #8636

@goldmerc Livewire controls the query string on the front end, so it's likely updating the URL each time an input is changed (that has URL). Hence why when you hit back, the previous change is after the first input change.

There is a URL replace option, but then I think it will just overwrite the URL completely and there won't be a back.

**Comment by goldmerc** [comment]
In the attached repo example, the url is only updated when the form is submitted. So, from the user experience point of view, on form submission the url changes from

`app.test/price-range-test`

to

`app.test/price-range-test?minPrice=20&maxPrice=50`

Which would be the expected behaviour. It's only when you hit back that the extra history step is revealed. Interestingly, it doesn't matter which order the minPrice & maxPrice inputs are set. Even if maxPrice is set first, the extra history step is 

`app.test/price-range-test?minPrice=20`

It seems to me that when livewire updates multiple component properties with #[Url(history: true)] in the same request (ie on submit), the query string update should be batched to include all the changes in one go. I think that is what the user would expect when they use a submit button to make multiple changes at once. 

I think I also found another query string issue (https://github.com/livewire/livewire/discussions/8640#discussioncomment-10020990). This is not so important for me but I could open another repo for it if you want?

**Comment by xshanj2** [comment]
@goldmerc Hi, I encountered the same problem as you, could you please tell me how you solved it?

**Reply by goldmerc** [reply]
@xshanj2 from memory, you can avoid the issue by using a property which is an array. So, instead of...

```
#[Url(history: true)]
public $minPrice;

#[Url(history: true)]
public $maxPrice;
```

you can do...

```
#[Url(history: true)]
public $price = [
    'min' => '',
    'max' => '',
];
```

Or you might prefer to call the array $filters for instance.

Hope it works!

**Reply by xshanj2** [reply]
That's a good idea, thank you!
